### PR TITLE
Implemented slack login

### PR DESCRIPTION
**Pull request by ethankeshishian**
User is prompted to log in when first accessing the website. After logging in with Slack, the user is redirected to the main homepage. 

Notes:
* Make sure to update the frontend with npm run build. 
* Make sure to update the .env file with the new slack variable.

### CODE REVIEW

File: react/src/components/Gallery/index.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import React, { Fragment } from "react";
+import RUG from "react-upload-gallery";
+import "react-upload-gallery/dist/style.css";
+import axios from 'axios';
+
+// import { initialState } from "./data";
+
+class Gallery extends React.Component {
```
Comment (DumboOctopus) [inline react/src/components/Gallery/index.js:8]: is this component still used?

File: django_project/django_project/settings.py [hunk 2]
```diff
@@ -30,6 +30,11 @@
     SECRET_KEY = os.environ.get('SECRET_KEY')
 else:
     SECRET_KEY = '_i@@y#sw+z^q1+xk2wbr2$cz1$-ll44(!f74lce!e$qkw$i(dg'
+if os.environ.get('SLACK_SECRET_KEY'):
+    SLACK_SECRET_KEY = os.environ.get('SLACK_SECRET_KEY')
+else:
```
Comment (DumboOctopus) [inline django_project/django_project/settings.py:35]: with this, you don't have to have the if statement because we don't really have a default value like we do for SECRET_KEY.  If the person doesn't provide a key, the system should just fail. So you could simplify it to
```
SLACK_SECRET_KEY = os.environ.get('SLACK_SECRET_KEY')
```

File: django_project/oauth/views.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import requests
+import json
+from django.contrib.auth.models import User
+from django.contrib.auth import authenticate, login
+import os
+import environ
+from django.contrib.sessions.models import Session
+from django_project.settings import SLACK_SECRET_KEY
+
+
```
[the remark this answers is not in the document]
Comment (DumboOctopus) [reply django_project/oauth/views.py:12]: And you can ask me for examples of how to log if need be

File: react/src/components/Home/index.js [hunk 4]
```diff
@@ -1,18 +1,27 @@
...
-            <div>Home Page</div>
-            <div>
-                <Link to='/create'>Create</Link>
-            </div>
-            <div>
-                <Link to='/update/1'>Update</Link>
-            </div>
-        </div>
-    );
+  console.log('props.isLoggedIn', props.isLoggedIn);
```
Comment (DumboOctopus) [inline react/src/components/Home/index.js:5]: be sure to remove these debugging statements as well

**Review by DumboOctopus — CHANGES_REQUESTED** [review]
Hi guys! Overall it looks pretty solid. I made some comments on things I think need some refactoring/cleanup. If you wanna discuss any of them, feel free to just message me on slack!

After we resolve all the comments in this review, you can push your modified branch to github and it will automatically update the pull request. Then I'll take another look at it and make sure to run and test some edge cases. If I find anything sus, I'll let you guys know and we can work it out. Otherwise, I'll approve the request and you guys can click the honrary green merge button

File: react/src/components/Gallery/index.js (hunk 1 above)
Comment (ethankeshishian) [reply react/src/components/Gallery/index.js:8]: Not sure, we didn't make any changes here.

File: django_project/oauth/views.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        # sessions.delete()
+        # instance.delete()
+        if user is not None:
+            # The user already exists in the database
+            # return HttpResponse("Success with an existing user")
+            login(request, user)
+            return redirect('/')
+        else:
+            # A new user needs to be made
+            new_user = User.objects.create_user(username=userId, password='')
```
[the remark this answers is not in the document]
Comment (ethankeshishian) [reply django_project/oauth/views.py:57]: Looks like slack returns the whole name as a string. Not sure if we can parse this if names aren't always one word for the first and one for the last name?

File: react/src/components/Gallery/index.js (hunk 1 above)
Comment (DumboOctopus) [reply react/src/components/Gallery/index.js:8]: hmm ok could you delete this file then

File: django_project/oauth/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (DumboOctopus) [reply django_project/oauth/views.py:58]: ohh I see. Hmm in that case we can just not do that for now

File: django_project/oauth/views.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    # data to be sent to api 
+    print(SLACK_SECRET_KEY)
+    data = {
+        'client_id':'4526132454.1654017102375',
+        'client_secret':SLACK_SECRET_KEY,
+        'code':code, 
+        'redirect_uri':'http://localhost:8000/django/oauth/slack_auth' 
+        } 
+    # sending post request and saving response as response object 
+    r = requests.post(url = API_ENDPOINT, data = data)
```
[the remark this answers is not in the document]
Comment (aritramullick) [reply django_project/oauth/views.py:31]: Added try-catch to this block.

File: django_project/oauth/views.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (aritramullick) [reply django_project/oauth/views.py:57]: So, I found a way to get the User's name. However, I think we might run into some issues if we try to save it as first name and last name (not all users will have exactly 2 words in their name - they could have middle names, or they could use only one nickname for Slack). So, for now, I am saving it as first_name in the User profile only. It should be easy to modify if we want to make changes later!

File: django_project/oauth/views.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            # return HttpResponse("Success after making a new user")
+            return redirect('/')
+    if not loggedIn:
+        return HttpResponse("Failure! ok=false")
+
+def view(request):
+    print(request.user)
+    return "anything"
+
+def get_logged_in(request):
```
[the remark this answers is not in the document]
Comment (aritramullick) [reply django_project/oauth/views.py:96]: I think this code has changed a bit, so I am unsure about what to do.

**Review by DumboOctopus — APPROVED** [review]
There just some dead code but besides that it looks good! So I'm gonna approve it. Please do remove these lines (just so the code stays clean). After thats done, feel free to merge!

File: react/src/components/App.js [hunk 8]
```diff
@@ -1,18 +1,45 @@
...
     <div className="App">
       <Switch>
         <Route path="/create" component={CreateUpdateGallery} />
         <Route path="/update/:id" component={CreateUpdateGallery} />
+        <Route path="/signup" component={Signup} />
+        <Route
+          path="/"
+          render={() => <Home isLoggedIn={loggedIn} isLoading={isLoading} />}
+        />
         <Route path="/rearrange" component={RearrangeImages} />
```
[the remark this answers is not in the document]
Comment (aritramullick) [reply react/src/components/App.js:42]: Solving this would just be moving the rearrange path above the '/' path?

File: react/package.json [hunk 9]
```diff
@@ -21,6 +21,8 @@
     "react-redux": "^7.2.2",
     "react-router-dom": "^5.2.0",
     "react-scripts": "3.3.1",
+    "react-slack-login": "^1.1.7",
+    "react-upload-gallery": "^0.2.0",
```
[the remark this answers is not in the document]
Comment (aritramullick) [reply react/package.json:25]: How so? Do we not still need react-slack-login when a new user tries to sign in?

[the remark this answers is not in the document]
Comment (DumboOctopus) [reply react/package.json:25]: I don't think so just because we are not using the slack react sign in button in the frontend anymore; instead we are using just an `a` tag i think.
